File: app/src/main/java/com/battlelancer/seriesguide/ui/stats/StatsFragment.java
package com.battlelancer.seriesguide.ui.stats;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.settings.DisplaySettings;
import com.battlelancer.seriesguide.util.ShareUtils;
import com.battlelancer.seriesguide.widgets.EmptyView;
import java.text.NumberFormat;
import java.util.Locale;

/**
 * Displays some statistics about the users show database, e.g. number of shows, episodes, share of
 * watched episodes, etc.
 */
public class StatsFragment extends Fragment {

    @BindView(R.id.emptyViewStats) EmptyView errorView;

    @BindView(R.id.textViewStatsShows) TextView textViewShows;
    @BindView(R.id.textViewStatsShowsWithNext) TextView textViewShowsWithNextEpisode;
    @BindView(R.id.progressBarStatsShowsWithNext) ProgressBar progressBarShowsWithNextEpisode;
    @BindView(R.id.textViewStatsShowsContinuing) TextView textViewShowsContinuing;
    @BindView(R.id.progressBarStatsShowsContinuing) ProgressBar progressBarShowsContinuing;

    @BindView(R.id.textViewStatsEpisodes) TextView textViewEpisodes;
    @BindView(R.id.textViewStatsEpisodesWatched) TextView textViewEpisodesWatched;
    @BindView(R.id.progressBarStatsEpisodesWatched) ProgressBar progressBarEpisodesWatched;
    @BindView(R.id.textViewStatsEpisodesRuntime) TextView textViewEpisodesRuntime;
    @BindView(R.id.progressBarStatsEpisodesRuntime) ProgressBar progressBarEpisodesRuntime;

    @BindView(R.id.textViewStatsMovies) TextView textViewMovies;
    @BindView(R.id.textViewStatsMoviesWatchlist) TextView textViewMoviesWatchlist;
    @BindView(R.id.progressBarStatsMoviesWatchlist) ProgressBar progressBarMoviesWatchlist;
    @BindView(R.id.textViewStatsMoviesWatchlistRuntime) TextView textViewMoviesWatchlistRuntime;

    private Unbinder unbinder;
    private StatsViewModel model;
    private StatsLiveData.Stats currentStats;
    private boolean hasFinalValues;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_stats, container, false);
        unbinder = ButterKnife.bind(this, v);

        errorView.setVisibility(View.GONE);
        errorView.setButtonClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loadStats();
            }
        });

        // set some views invisible so they can be animated in once stats are computed
        textViewShowsWithNextEpisode.setVisibility(View.INVISIBLE);
        progressBarShowsWithNextEpisode.setVisibility(View.INVISIBLE);
        textViewShowsContinuing.setVisibility(View.INVISIBLE);
        progressBarShowsContinuing.setVisibility(View.INVISIBLE);

        textViewEpisodesWatched.setVisibility(View.INVISIBLE);
        progressBarEpisodesWatched.setVisibility(View.INVISIBLE);
        textViewEpisodesRuntime.setVisibility(View.INVISIBLE);

        textViewMoviesWatchlist.setVisibility(View.INVISIBLE);
        progressBarMoviesWatchlist.setVisibility(View.INVISIBLE);
        textViewMoviesWatchlistRuntime.setVisibility(View.INVISIBLE);

        return v;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setHasOptionsMenu(true);

        model = ViewModelProviders.of(this).get(StatsViewModel.class);
        model.getStatsData().observe(this, new Observer<StatsLiveData.StatsUpdateEvent>() {
            @Override
            public void onChanged(@Nullable StatsLiveData.StatsUpdateEvent statsUpdateEvent) {
                handleStatsUpdate(statsUpdateEvent);
            }
        });
        loadStats();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        unbinder.unbind();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);

        // guard against not attached to activity
        if (!isAdded()) {
            return;
        }

        inflater.inflate(R.menu.stats_menu, menu);

        menu.findItem(R.id.menu_action_stats_filter_specials)
                .setChecked(DisplaySettings.isHidingSpecials(getActivity()));
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int itemId = item.getItemId();
        if (itemId == R.id.menu_action_stats_share) {
            shareStats();
            return true;
        }
        if (itemId == R.id.menu_action_stats_filter_specials) {
            PreferenceManager.getDefaultSharedPreferences(getActivity()).edit()
                    .putBoolean(DisplaySettings.KEY_HIDE_SPECIALS, !item.isChecked())
                    .apply();
            //noinspection ConstantConditions always attached to activity
            getActivity().invalidateOptionsMenu();
            loadStats();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void loadStats() {
        model.getStatsData().loadStats();
    }

    private void handleStatsUpdate(StatsLiveData.StatsUpdateEvent event) {
        if (!isAdded()) {
            return;
        }
        currentStats = event.stats;
        hasFinalValues = event.finalValues;
        updateStats(event.stats, event.finalValues, event.successful);
    }

    private void updateStats(@NonNull StatsLiveData.Stats stats, boolean hasFinalValues,
            boolean successful) {
        // display error if not all stats could be calculated
        errorView.setVisibility(successful ? View.GONE : View.VISIBLE);

        NumberFormat format = NumberFormat.getIntegerInstance();

        // all shows
        textViewShows.setText(format.format(stats.shows));

        // shows with next episodes
        progressBarShowsWithNextEpisode.setMax(stats.shows);
        progressBarShowsWithNextEpisode.setProgress(stats.showsWithNextEpisodes);
        progressBarShowsWithNextEpisode.setVisibility(View.VISIBLE);

        textViewShowsWithNextEpisode.setText(getString(R.string.shows_with_next,
                format.format(stats.showsWithNextEpisodes)).toUpperCase(Locale.getDefault()));
        textViewShowsWithNextEpisode.setVisibility(View.VISIBLE);

        // continuing shows
        progressBarShowsContinuing.setMax(stats.shows);
        progressBarShowsContinuing.setProgress(stats.showsContinuing);
        progressBarShowsContinuing.setVisibility(View.VISIBLE);

        textViewShowsContinuing.setText(getString(R.string.shows_continuing,
                format.format(stats.showsContinuing)).toUpperCase(Locale.getDefault()));
        textViewShowsContinuing.setVisibility(View.VISIBLE);

        // all episodes
        textViewEpisodes.setText(format.format(stats.episodes));

        // watched episodes
        progressBarEpisodesWatched.setMax(stats.episodes);
        progressBarEpisodesWatched.setProgress(stats.episodesWatched);
        progressBarEpisodesWatched.setVisibility(View.VISIBLE);

        textViewEpisodesWatched.setText(getString(R.string.episodes_watched,
                format.format(stats.episodesWatched)).toUpperCase(Locale.getDefault()));
        textViewEpisodesWatched.setVisibility(View.VISIBLE);

        // episode runtime
        String watchedDuration = getTimeDuration(stats.episodesWatchedRuntime);
        if (!hasFinalValues) {
            // showing minimum (= not the final value)
            watchedDuration = "> " + watchedDuration;
        }
        textViewEpisodesRuntime.setText(watchedDuration);
        textViewEpisodesRuntime.setVisibility(View.VISIBLE);
        progressBarEpisodesRuntime.setVisibility(successful ?
                (hasFinalValues ? View.GONE : View.VISIBLE)
                : View.GONE);

        // movies
        textViewMovies.setText(format.format(stats.movies));

        // movies in watchlist
        progressBarMoviesWatchlist.setMax(stats.movies);
        progressBarMoviesWatchlist.setProgress(stats.moviesWatchlist);
        progressBarMoviesWatchlist.setVisibility(View.VISIBLE);

        textViewMoviesWatchlist.setText(getString(R.string.movies_on_watchlist,
                format.format(stats.moviesWatchlist)).toUpperCase(Locale.getDefault()));
        textViewMoviesWatchlist.setVisibility(View.VISIBLE);

        // runtime of movie watchlist
        textViewMoviesWatchlistRuntime.setText(getTimeDuration(stats.moviesWatchlistRuntime));
        textViewMoviesWatchlistRuntime.setVisibility(View.VISIBLE);
    }

    private String getTimeDuration(long duration) {
        long days = duration / DateUtils.DAY_IN_MILLIS;
        duration %= DateUtils.DAY_IN_MILLIS;
        long hours = duration / DateUtils.HOUR_IN_MILLIS;
        duration %= DateUtils.HOUR_IN_MILLIS;
        long minutes = duration / DateUtils.MINUTE_IN_MILLIS;

        StringBuilder result = new StringBuilder();
        if (days != 0) {
            result.append(getResources().getQuantityString(R.plurals.days_plural, (int) days,
                    (int) days));
        }
        if (hours != 0) {
            if (days != 0) {
                result.append(" ");
            }
            result.append(getResources().getQuantityString(R.plurals.hours_plural, (int) hours,
                    (int) hours));
        }
        if (minutes != 0 || (days == 0 && hours == 0)) {
            if (days != 0 || hours != 0) {
                result.append(" ");
            }
            result.append(getResources().getQuantityString(R.plurals.minutes_plural,
                    (int) minutes,
                    (int) minutes));
        }

        return result.toString();
    }

    private void shareStats() {
        if (currentStats == null) {
            return;
        }

        NumberFormat format = NumberFormat.getIntegerInstance();

        StringBuilder statsString = new StringBuilder();
        statsString.append(getString(R.string.app_name))
                .append(" ")
                .append(getString(R.string.statistics));
        statsString.append("\n");
        statsString.append("\n");
        // shows
        statsString.append(format.format(currentStats.shows))
                .append(" ")
                .append(getString(R.string.statistics_shows));
        statsString.append("\n");
        statsString.append(getString(R.string.shows_with_next,
                format.format(currentStats.showsWithNextEpisodes)));
        statsString.append("\n");
        statsString.append(getString(R.string.shows_continuing,
                format.format(currentStats.showsContinuing)));
        statsString.append("\n");
        statsString.append("\n");
        // episodes
        statsString.append(format.format(currentStats.episodes)).append(" ").append(
                getString(R.string.statistics_episodes));
        statsString.append("\n");
        statsString.append(getString(R.string.episodes_watched,
                format.format(currentStats.episodesWatched)));
        statsString.append("\n");
        if (currentStats.episodesWatchedRuntime != 0) {
            String watchedDuration = getTimeDuration(currentStats.episodesWatchedRuntime);
            if (!hasFinalValues) {
                // showing minimum (= not the final value)
                watchedDuration = "> " + watchedDuration;
            }
            statsString.append(watchedDuration)
                    .append(" ")
                    .append(getString(R.string.runtime_all_episodes));
            statsString.append("\n");
        }
        statsString.append("\n");
        // movies
        statsString.append(format.format(currentStats.movies))
                .append(" ")
                .append(getString(R.string.statistics_movies));
        statsString.append("\n");
        statsString.append(getString(R.string.movies_on_watchlist,
                format.format(currentStats.moviesWatchlist)));
        statsString.append("\n");
        statsString.append(getTimeDuration(currentStats.moviesWatchlistRuntime))
                .append(" ")
                .append(getString(R.string.runtime_movies_watchlist));

        ShareUtils.startShareIntentChooser(getActivity(), statsString.toString(), R.string.share);
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/ui/stats/StatsFragment.kt
package com.battlelancer.seriesguide.ui.stats

import android.os.Bundle
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isGone
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.preference.PreferenceManager
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.databinding.FragmentStatsBinding
import com.battlelancer.seriesguide.settings.DisplaySettings
import com.battlelancer.seriesguide.util.ShareUtils
import com.battlelancer.seriesguide.util.copyTextToClipboardOnLongClick
import java.text.NumberFormat

/**
 * Displays some statistics about the users show database, e.g. number of shows, episodes, share of
 * watched episodes, etc.
 */
class StatsFragment : Fragment() {

    private var binding: FragmentStatsBinding? = null
    private val model by viewModels<StatsViewModel>()
    private var currentStats: Stats? = null
    private var hasFinalValues: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentStatsBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val binding = binding!!
        binding.errorView.visibility = View.GONE
        binding.errorView.setButtonClickListener { loadStats() }

        // set some views invisible so they can be animated in once stats are computed
        binding.textViewShowsFinished.visibility = View.INVISIBLE
        binding.progressBarShowsFinished.visibility = View.INVISIBLE
        binding.textViewShowsWithNextEpisode.visibility = View.INVISIBLE
        binding.progressBarShowsWithNextEpisode.visibility = View.INVISIBLE
        binding.textViewShowsContinuing.visibility = View.INVISIBLE
        binding.progressBarShowsContinuing.visibility = View.INVISIBLE

        binding.textViewEpisodesWatched.visibility = View.INVISIBLE
        binding.progressBarEpisodesWatched.visibility = View.INVISIBLE
        binding.textViewEpisodesRuntime.visibility = View.INVISIBLE

        binding.textViewMoviesWatchlist.visibility = View.INVISIBLE
        binding.textViewMoviesWatchlistRuntime.visibility = View.INVISIBLE

        binding.progressBarMoviesWatched.visibility = View.INVISIBLE
        binding.textViewMoviesWatched.visibility = View.INVISIBLE
        binding.textViewMoviesWatchedRuntime.visibility = View.INVISIBLE

        binding.progressBarMoviesCollection.visibility = View.INVISIBLE
        binding.textViewMoviesCollection.visibility = View.INVISIBLE
        binding.textViewMoviesCollectionRuntime.visibility = View.INVISIBLE

        // set up long-press to copy text to clipboard (d-pad friendly vs text selection)
        binding.textViewShows.copyTextToClipboardOnLongClick()
        binding.textViewShowsFinished.copyTextToClipboardOnLongClick()
        binding.textViewShowsWithNextEpisode.copyTextToClipboardOnLongClick()
        binding.textViewShowsContinuing.copyTextToClipboardOnLongClick()
        binding.textViewEpisodes.copyTextToClipboardOnLongClick()
        binding.textViewEpisodesWatched.copyTextToClipboardOnLongClick()
        binding.textViewEpisodesRuntime.copyTextToClipboardOnLongClick()
        binding.textViewMovies.copyTextToClipboardOnLongClick()
        binding.textViewMoviesWatchlist.copyTextToClipboardOnLongClick()
        binding.textViewMoviesWatchlistRuntime.copyTextToClipboardOnLongClick()
        binding.textViewMoviesWatched.copyTextToClipboardOnLongClick()
        binding.textViewMoviesWatchedRuntime.copyTextToClipboardOnLongClick()
        binding.textViewMoviesCollection.copyTextToClipboardOnLongClick()
        binding.textViewMoviesCollectionRuntime.copyTextToClipboardOnLongClick()

        model.statsData.observe(viewLifecycleOwner, { this.handleStatsUpdate(it) })
    }

    override fun onDestroyView() {
        super.onDestroyView()

        binding = null
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)

        // guard against not attached to activity
        if (!isAdded) {
            return
        }

        inflater.inflate(R.menu.stats_menu, menu)

        menu.findItem(R.id.menu_action_stats_filter_specials).isChecked =
            DisplaySettings.isHidingSpecials(requireContext())
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId = item.itemId
        if (itemId == R.id.menu_action_stats_share) {
            shareStats()
            return true
        }
        if (itemId == R.id.menu_action_stats_filter_specials) {
            PreferenceManager.getDefaultSharedPreferences(activity).edit()
                .putBoolean(DisplaySettings.KEY_HIDE_SPECIALS, !item.isChecked)
                .apply()

            requireActivity().invalidateOptionsMenu()
            loadStats()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun loadStats() {
        model.hideSpecials.value = DisplaySettings.isHidingSpecials(requireContext())
    }

    private fun handleStatsUpdate(event: StatsUpdateEvent) {
        if (!isAdded) {
            return
        }
        currentStats = event.stats
        hasFinalValues = event.finalValues
        updateStats(event.stats, event.finalValues, event.successful)
    }

    private fun updateStats(
        stats: Stats, hasFinalValues: Boolean,
        successful: Boolean
    ) {
        val binding = binding ?: return

        // display error if not all stats could be calculated
        binding.errorView.isGone = successful

        val format = NumberFormat.getIntegerInstance()

        // all shows
        binding.textViewShows.text = format.format(stats.shows.toLong())

        // shows finished
        binding.progressBarShowsFinished.apply {
            max = stats.shows
            progress = stats.showsFinished
            visibility = View.VISIBLE
        }

        binding.textViewShowsFinished.apply {
            text = getString(
                R.string.shows_finished,
                format.format(stats.showsFinished.toLong())
            )
            visibility = View.VISIBLE
        }

        // shows with next episodes
        binding.progressBarShowsWithNextEpisode.apply {
            max = stats.shows
            progress = stats.showsWithNextEpisodes
            visibility = View.VISIBLE
        }

        binding.textViewShowsWithNextEpisode.apply {
            text = getString(
                R.string.shows_with_next,
                format.format(stats.showsWithNextEpisodes.toLong())
            )
            visibility = View.VISIBLE
        }

        // continuing shows
        binding.progressBarShowsContinuing.apply {
            max = stats.shows
            progress = stats.showsContinuing
            visibility = View.VISIBLE
        }

        binding.textViewShowsContinuing.text = getString(
            R.string.shows_continuing,
            format.format(stats.showsContinuing.toLong())
        )
        binding.textViewShowsContinuing.visibility = View.VISIBLE

        // all episodes
        binding.textViewEpisodes.text = format.format(stats.episodes.toLong())

        // watched episodes
        binding.progressBarEpisodesWatched.max = stats.episodes
        binding.progressBarEpisodesWatched.progress = stats.episodesWatched
        binding.progressBarEpisodesWatched.visibility = View.VISIBLE

        binding.textViewEpisodesWatched.text = getString(
            R.string.episodes_watched,
            format.format(stats.episodesWatched.toLong())
        )
        binding.textViewEpisodesWatched.visibility = View.VISIBLE

        // episode runtime
        var watchedDuration = getTimeDuration(stats.episodesWatchedRuntime)
        if (!hasFinalValues) {
            // showing minimum (= not the final value)
            watchedDuration = "> $watchedDuration"
        }
        binding.textViewEpisodesRuntime.text = watchedDuration
        binding.textViewEpisodesRuntime.visibility = View.VISIBLE
        binding.progressBarEpisodesRuntime.visibility = if (successful)
            if (hasFinalValues) View.GONE else View.VISIBLE
        else
            View.GONE

        // movies
        binding.textViewMovies.text = format.format(stats.movies.toLong())

        // watched movies
        binding.progressBarMoviesWatched.apply {
            max = stats.movies
            progress = stats.moviesWatched
            visibility = View.VISIBLE
        }
        binding.textViewMoviesWatched.apply {
            text = getString(
                R.string.movies_watched_format,
                format.format(stats.moviesWatched.toLong())
            )
            visibility = View.VISIBLE
        }
        binding.textViewMoviesWatchedRuntime.apply {
            text = getTimeDuration(stats.moviesWatchedRuntime)
            visibility = View.VISIBLE
        }

        // movies in watchlist
        binding.textViewMoviesWatchlist.apply {
            text = getString(
                R.string.movies_on_watchlist,
                format.format(stats.moviesWatchlist.toLong())
            )
            visibility = View.VISIBLE
        }
        binding.textViewMoviesWatchlistRuntime.apply {
            text = getTimeDuration(stats.moviesWatchlistRuntime)
            visibility = View.VISIBLE
        }

        // movies in collection
        binding.progressBarMoviesCollection.apply {
            max = stats.movies
            progress = stats.moviesCollection
            visibility = View.VISIBLE
        }
        binding.textViewMoviesCollection.apply {
            text = getString(R.string.stats_in_collection_format, stats.moviesCollection)
            visibility = View.VISIBLE
        }
        binding.textViewMoviesCollectionRuntime.apply {
            text = getTimeDuration(stats.moviesCollectionRuntime)
            visibility = View.VISIBLE
        }
    }

    private fun getTimeDuration(duration: Long): String {
        var durationCalc = duration
        val days = durationCalc / DateUtils.DAY_IN_MILLIS
        durationCalc %= DateUtils.DAY_IN_MILLIS
        val hours = durationCalc / DateUtils.HOUR_IN_MILLIS
        durationCalc %= DateUtils.HOUR_IN_MILLIS
        val minutes = durationCalc / DateUtils.MINUTE_IN_MILLIS

        val result = StringBuilder()
        if (days != 0L) {
            result.append(
                resources.getQuantityString(
                    R.plurals.days_plural, days.toInt(),
                    days.toInt()
                )
            )
        }
        if (hours != 0L) {
            if (days != 0L) {
                result.append(" ")
            }
            result.append(
                resources.getQuantityString(
                    R.plurals.hours_plural, hours.toInt(),
                    hours.toInt()
                )
            )
        }
        if (minutes != 0L || days == 0L && hours == 0L) {
            if (days != 0L || hours != 0L) {
                result.append(" ")
            }
            result.append(
                resources.getQuantityString(
                    R.plurals.minutes_plural,
                    minutes.toInt(),
                    minutes.toInt()
                )
            )
        }

        return result.toString()
    }

    private fun shareStats() {
        val currentStats = this.currentStats ?: return

        val format = NumberFormat.getIntegerInstance()

        val shows = format.format(currentStats.shows.toLong())
        val showsWithNext = getString(
            R.string.shows_with_next,
            format.format(currentStats.showsWithNextEpisodes.toLong())
        )
        val showsContinuing = getString(
            R.string.shows_continuing,
            format.format(currentStats.showsContinuing.toLong())
        )
        val showsFinished = getString(
            R.string.shows_finished,
            format.format(currentStats.showsFinished.toLong())
        )

        val episodes = format.format(currentStats.episodes.toLong())
        val episodesWatched = getString(
            R.string.episodes_watched,
            format.format(currentStats.episodesWatched.toLong())
        )

        val showStats =
            "${getString(R.string.app_name)} ${getString(R.string.statistics)}\n\n" +
                    "${getString(R.string.shows)}\n" +
                    "$shows\n" +
                    "$showsWithNext\n" +
                    "$showsContinuing\n" +
                    "$showsFinished\n\n" +
                    "${getString(R.string.episodes)}\n" +
                    "$episodes\n" +
                    "$episodesWatched\n"

        val statsString = StringBuilder(showStats)
        if (currentStats.episodesWatchedRuntime != 0L) {
            var watchedDuration = getTimeDuration(currentStats.episodesWatchedRuntime)
            if (!hasFinalValues) {
                // showing minimum (= not the final value)
                watchedDuration = "> $watchedDuration"
            }
            statsString.append("$watchedDuration\n")
        }

        statsString.append("\n")

        // movies
        val movies = format.format(currentStats.movies.toLong())
        val moviesWatchlist = getString(
            R.string.movies_on_watchlist,
            format.format(currentStats.moviesWatchlist.toLong())
        )
        val moviesWatched = getString(
            R.string.movies_watched_format,
            format.format(currentStats.moviesWatched.toLong())
        )
        val moviesCollection = getString(
            R.string.stats_in_collection_format,
            currentStats.moviesCollection
        )
        val moviesWatchlistRuntime = getTimeDuration(currentStats.moviesWatchlistRuntime)
        val moviesWatchedRuntime = getTimeDuration(currentStats.moviesWatchedRuntime)
        val moviesCollectionRuntime = getTimeDuration(currentStats.moviesCollectionRuntime)

        val movieStats = "${getString(R.string.movies)}\n" +
                "$movies\n" +
                "$moviesWatched\n" +
                "$moviesWatchedRuntime\n" +
                "$moviesWatchlist\n" +
                "$moviesWatchlistRuntime\n" +
                "$moviesCollection\n" +
                "$moviesCollectionRuntime\n"

        statsString.append(movieStats)

        ShareUtils.startShareIntentChooser(activity, statsString.toString(), R.string.share)
    }
}
